### DISCUSSION TITLE
StateNotifier lost state when parent provider refresh

### DISCUSSION THREAD

**Discussion by PandaGeek1024** [rrousselGit/riverpod#1176] [Q&A]
I have been trying to work out a solution for it but no luck. The key problem here is when you have a StateNotifier that depends on another provider, A fresh from the parent provider will cause the StateNotifier to be regenerated and result in existing state lost. I will use a simple example to explains my problem.
Here are the providers:
```
class Pair {
  final int parentValue;
  final int childValue;

  Pair({required this.parentValue, required this.childValue});

  Pair copyWith({int? parentValue, int? childValue}) {
    return Pair(
        parentValue: parentValue ?? this.parentValue,
        childValue: childValue ?? this.childValue
    );
  }
}

class PairNotifier extends StateNotifier<Pair> {
  // parentValue comes from counterProvider but childValue is maintained in existing notifier
  PairNotifier(int parentValue): super(Pair(parentValue: parentValue, childValue: 0));

  Pair increment() {
    final value = state.childValue + 1;
    state = state.copyWith(childValue: value);
    return state;
  }
}

final counterProvider = StateProvider((ref) => 0);

final pairStateProvider = StateNotifierProvider<PairNotifier, Pair>((ref) {
  final parentValue = ref.watch(counterProvider); 
  // when counter is updated, PairNotifier is regenerated therefore childValue is lost.
  return PairNotifier(parentValue);
});
```
A real world case might be quite differentIs but this should be enough to deliver the idea. I would to know if there is a way to preserve the existing state when all its' parent providers refreshed.

Remaining code:
```
void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ProviderScope(
      child: MaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: const MyHomePage(title: 'Flutter Demo Home Page'),
      ),
    );
  }
}

class MyHomePage extends ConsumerStatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends ConsumerState<MyHomePage> {

  void _incrementCounter() {
    // counterProvider updated cause refresh.
    ref.read(counterProvider.notifier).state++;
  }

  @override
  Widget build(BuildContext context) {
    final pair = ref.watch(pairStateProvider);
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              '${pair.childValue}',
              style: Theme.of(context).textTheme.headline4,
            ),
            ElevatedButton(onPressed: () {
              ref.read(pairStateProvider.notifier).increment();
            }, child: Text("Click to update childValue"),),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}
```

**Comment by rrousselGit** [comment]
That's expected. It's part of how `ref.watch` works


If you don't want that, don't use `ref.watch` and instead use `ref.listen`

**Comment by PandaGeek1024** [comment]
Thanks @rrousselGit  for your quick response. 
I'm not quite sure how I can achieve it with ref.listen, could you please provide an example?

**Comment by PandaGeek1024** [comment]
Isn't it a good idea to return the current state with ref? So we can update the state base on previous value.
StateNotifierProvider<Counter, int>((ref, state) {});

**Comment by PandaGeek1024** [comment]
@rrousselGit ref.listen simply doesn't work in my case especially when I need to watch multiple providers . Is there a better solution? The current workaround that I can think of is to have the state in a top provider and have another viewmodel provider to watch this state provider. Something like:
```
final counterProvider = StateProvider((ref) => 0);

final stateProvider = StateProvider((ref) => Pair(0,0));

class ViewModel {
  // parentValue comes from counterProvider but childValue is maintained in existing notifier
  ViewModel(Pair state, StateController controller)

  void update() {
    controller.state = Pair(1,1);
 }

final pairStateProvider = StateProvider<Pair>((ref) {
  final parentValue = ref.watch(counterProvider); 
  final controller = ref.watch(stateProvider.notifier); 
  return ViewModel(controller.state.copyWith(parentValue: parentValue), controller);
});
```
But honestly I really don't want to do it this way because that would mean I need to give up StateNotifier which should be a great class to use.

**Comment by rrousselGit** [comment]
You can listen to multiple providers too. I don't see the issue

**Comment by PandaGeek1024** [comment]
As usual, thanks for your reply. I did try with listen but really don't seem to work for me. 
Please correct me if Im wrong.
```
final pairStateProvider = StateNotifierProvider<PairNotifier, Pair>((ref) {
  late int parentValue;
  ref.listen<int>(counterProvider, (previous, next) {
    parentValue = next;
  }, fireImmediately: true,);
  return PairNotifier(parentValue);
});
```
In that case, the PairNotifier is not regenerated but we also don't get the updated parentValue.

**Comment by rrousselGit** [comment]
It's your responsibility to update the PairNotifier when the counter changes

You can do:

```dart
final notifier = PairNotifier(...);
ref.listen<int>(counterProvider, (previous, next) {
  notifier.pairChanged(previous, next); // have PairNotifier handle updates
});
return notifier;
```

**Comment by PandaGeek1024** [comment]
Hmm.. interesting. Just tried it but seems quite verbose in the (common)case that if I have multiple providers to watch.
```
final pairStateProvider = StateNotifierProvider<PairNotifier, Pair>((ref) {
  final notifier = PairNotifier();//And can not initialize with values from counterProvider and etc.
  ref.listen<int>(counterProvider, (previous, next) {
    notifier.pairChanged(previous, next);
  });
  ref.listen<int>(authRepoProvider, (previous, next) {
    notifier.authRepoChanged(previous, next);
  });
  ref.listen<int>(cardsRepoProvider, (previous, next) {
    notifier.cardsRepoChanged(previous, next);
  });
  return notifier;
});
```
Moreover, If I need to run it for unit test, that means I would need to have another constructor or call those methods again to inject the authRepoProvider, cardsRepoProvider and etc. And It doesn't seem ideal.

It just feels weird that, the provider definitely cache the state but we are just not able to retrieve it.

I would really like to know how others deal with this elegantly.

**Comment by rrousselGit** [comment]
> Moreover, If I need to run it for unit test, that means I would need to have another constructor or call those methods again to inject the authRepoProvider, cardsRepoProvider and etc. And It doesn't seem ideal.



Your tests would typically read the provider instead of manually instantiating the StateNotifier, so the ref.listen calls would be executed.
After-all, you'd want to test them too

**Comment by PandaGeek1024** [comment]
I think if we can use ref.notifier.state to get the state, that will make our life much easier. And I think it totally make sense because we should be able to get the cached value. 
I can see that it is now prohibited to do it
```
  /// The [StateNotifier] currently exposed by this provider.
  ///
  /// Cannot be accessed while creating the provider.
  Notifier get notifier;
```
But maybe we can simply returns a null value if it is not created yet?

**Comment by rrousselGit** [comment]
I don't get what you're trying to do here.

You don't need ref.notifier to access the notifier as I shown before

**Comment by PandaGeek1024** [comment]
Sorry for the confusion. What I mean is if the provider cache the current state, then there should be a way to retrieve it?
So I'm thinking if this is workable?
```
final pairStateProvider = StateNotifierProvider<PairNotifier, Pair>((ref) {
  final parentValue = ref.watch(counterProvider); 
  final currentPairState = ref.notifier?.state //notifier return null if state is not created yet
  return PairNotifier(currentPairState ?? Pair(parentValue : parentValue, childValue : 0);
});
```

To me, this seems more elegant.
